### PR TITLE
Support for Rails3 session cookie (along with tests and docs)

### PR DESCRIPTION
**Pull request by saurabhnanda**
- It’s not possible to maintain parity with the Rails4 API because of the nature of the Rails3 cookie itself.
- decodeMaybe has, deliberately, not been added to promote usage of decodeEither. If the decoding fails, the reason for failure is very important and is the first thing that one looks for in the logs.
- Added a new function called `extractEither` which help extract a named cookie from the WAI request.
- Also added a helper function called `extactAndDecodeEither` to make the most common use-case a single function call. **Please note:** This adds a dependency on `wai`, but almost any user of this library is anyways already depending on WAI, so it really shouldn't matter.
- There is also `lookupUserIds` which is a convenience function to lookup the current UserId from the cookie, inserted by warden (most common authentication gem in Ruby/Rails)

### CODE REVIEW

**Comment by adinapoli** [comment]
Thank you @saurabhnanda ! For some reason I have completely missed this (reason being I receive emails only for @ mentions, hehe).

I will do a quick code review now and will leave the last word to @filib , but the fact you made the (right) choice of going for a separate module means I feel much lighter merging this in the virtue of the fact it won't break existing code 😉 

Alfredo

File: src/Web/Rails3/Session.hs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+-- | Lookup the Warden\/Devise UserIds from a decoded cookie. __Please note,__ a
+-- cookie may contain multiple UserIds, because it /seems/ that it is possible
+-- to be logged-in as multiple users simultaneously, if you define [multiple
+-- user
+-- models](https://github.com/plataformatec/devise/wiki/How-to-Setup-Multiple-Devise-User-Models)
+-- (the underlying data-structure allows it, as well).
+lookupUserIds :: (Num a) => RubyObject -> Maybe (NonEmpty a)
+lookupUserIds robj =
+  lookupKey ("warden.user.user.key", UTF_8) robj
+  >>= (\x -> fromRuby x :: Maybe [RubyObject]) -- [[int, int int], "random string"]
```
Comment (adinapoli) [inline src/Web/Rails3/Session.hs:137]: Wouldn't this be cleaner in do notation?

File: rails-session.cabal [hunk 2]
```diff
@@ -40,6 +48,7 @@ test-suite specs
   build-depends:       base
                      , bytestring
                      , filepath
+                     , semigroups
```
Comment (adinapoli) [inline rails-session.cabal:49]: If all you are using `semigroup` for is for `NonEmpty`, I think we can either add some `CPP` pragmas or decide whether or not we want to support older GHCs, as `base-4.6` (I think) onwards ships with `Data.List.NonEmpty` already. Not super fussed, though.

File: src/Web/Rails3/Session.hs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    compareChecksum checksum = if (computedChecksum == checksum) then (Right $ B64.decodeLenient b64) else (Left "[Rails3 Cookie] Checksum doesn't match")
+
+    computedChecksum :: Digest SHA1
+    computedChecksum = HMAC.hmacGetDigest (HMAC.hmac cookieSecret b64 :: HMAC SHA1)
+
+    (b64, hexChecksum) = let (a, b) = (breakSubstring delimiter $ urlDecode False x)
+                         in (a, BS.drop (BS.length delimiter) b)
+    delimiter = "--"
+
+    convertLeftToText e = case e of
```
Comment (adinapoli) [inline src/Web/Rails3/Session.hs]: Why not `either T.pack id` ?

File: rails-session.cabal (hunk 2 above)
Comment (saurabhnanda) [reply rails-session.cabal:49]: Will try to make this conditional. Does the repo have CI setup to test across different versions?

File: src/Web/Rails3/Session.hs (hunk 3 above)
Comment (saurabhnanda) [reply src/Web/Rails3/Session.hs]: Will make the change.

File: src/Web/Rails3/Session.hs (hunk 1 above)
Comment (saurabhnanda) [reply src/Web/Rails3/Session.hs:137]: Will make the change.

**Comment by saurabhnanda** [comment]
Any comments from @filib?

**Comment by philipcunningham** [comment]
Before merging, I would like to see:

* Remove of `wai` dependency - if someone is using a different stack, e.g. Snap, they shouldn't be expected to pull in `wai`. This could be extracted into another library e.g. `wai-rails-session`.
* Expose `decode` and `decodeEither` just like the Rails 4 API. It's up to consumers to decide how much information they need. This is broadly consistent with the API design of popular libraries like `aeson`.
* Remove marshalling of `String -> Text`. Whilst it's a good idea to use `Text`, this change first belongs upstream in `ruby-marshal`.

File: rails-session.cabal (hunk 2 above)
Comment (philipcunningham) [reply rails-session.cabal:49]: ~~Not at present https://github.com/iconnect/rails-session/issues/3~~

Comment (philipcunningham) [reply rails-session.cabal:49]: There's now CI and a build matrix that tests against GHC 7.10 and 8.0. 

If you pull in `master` it should work against your PR.

File: src/Web/Rails3/Session.hs (hunk 1 above)
Comment (saurabhnanda) [reply src/Web/Rails3/Session.hs:137]: Unable to make this change sensibly. Giving each intermediate value a sensible name is harder than I thought. If it's not too much of trouble, can we let this be as is?

File: rails-session.cabal (hunk 2 above)
Comment (saurabhnanda) [reply rails-session.cabal:49]: Thanks @filib .

**Comment by saurabhnanda** [comment]
@filib @adinapoli please take a look at the PR now. How do I trigger CI for this?

**Comment by adinapoli** [comment]
LGTM, but I'm keen on waiting for @filib 's last word as he was the one which originally requested for the provided changes. Thanks!

**Comment by philipcunningham** [comment]
https://github.com/iconnect/rails-session/issues/12

**Pull request merged**